Control/Monad/Accum.hs

```diff
@@ -0,0 +1,247 @@
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+-- | Module: Control.Monad.Accum
+-- Copyright: (C) Koz Ross 2022
+-- License: BSD-3-Clause (see the LICENSE file)
+-- Maintainer: koz.ross@retro-freedom.nz
+-- Stability: Experimental
+-- Portability: GHC only
+--
+-- [Computation type:] Accumulation (either append-only state, or writer with
+-- the ability to read all previous input).
+--
+-- [Binding strategy:] Binding a function to a monadic value monoidally
+-- accumulates the subcomputations (that is, using '<>').
+--
+-- [Useful for:] Logging, patch-style tracking.
+--
+-- [Zero and plus:] None.
+--
+-- [Example type:] @'Accum' w a@
+module Control.Monad.Accum
+  ( -- * Type class
+    MonadAccum (..),
+
+    -- * Lifting helper type
+    LiftingAccum (..),
+
+    -- * Other functions
+    looks,
+  )
+where
+
+import Control.Monad.Trans.Accum (AccumT)
+import qualified Control.Monad.Trans.Accum as Accum
+import Control.Monad.Trans.Class (MonadTrans (lift))
+import Control.Monad.Trans.Cont (ContT)
+import Control.Monad.Trans.Except (ExceptT)
+import Control.Monad.Trans.Identity (IdentityT)
+import Control.Monad.Trans.Maybe (MaybeT)
+import qualified Control.Monad.Trans.RWS.CPS as CPSRWS
+import qualified Control.Monad.Trans.RWS.Lazy as LazyRWS
+import qualified Control.Monad.Trans.RWS.Strict as StrictRWS
+import Control.Monad.Trans.Reader (ReaderT)
+import Control.Monad.Trans.Select (SelectT)
+import qualified Control.Monad.Trans.State.Lazy as LazyState
+import qualified Control.Monad.Trans.State.Strict as StrictState
+import qualified Control.Monad.Trans.Writer.CPS as CPSWriter
+import qualified Control.Monad.Trans.Writer.Lazy as LazyWriter
+import qualified Control.Monad.Trans.Writer.Strict as StrictWriter
+import Data.Functor (($>))
+import Data.Functor.Identity (Identity)
+import Data.Kind (Type)
+
+-- | The capability to accumulate. This can be seen in one of two ways:
+--
+-- * A 'MonadState' which can only append (using '<>'); or
+-- * A 'MonadWriter' (limited to 'tell') with the ability to view the result of
+-- all previous 'tell's.
+--
+-- = Laws
+--
+-- 'accum' should obey the following:
+--
+-- 1. @'accum' ('const' (x, 'mempty'))@ @=@ @'pure' x@
+-- 2. @'accum' f '*>' 'accum' g@ @=@
+-- @'accum' '$' \acc -> let (_, v) = f acc
+--                          (res, w) = g (acc '<>' v) in (res, v '<>' w)@
```

Comment:
I would name these laws: 

1.  `[Pure-Quot]` (in the sense of `pure x` being quotiented with `(x, mempty)`)
2. `[Const-Accum]` because it is what it says it is.

Reply:
We'd need to name all the others too, then. Good idea though.